exec/java-exec/src/main/java/org/apache/drill/exec/client/DrillClient.java

```diff
@@ -74,73 +74,148 @@
-    throws OutOfMemoryException {
+  /**
+   * @deprecated Create a DrillClient using {@link DrillClient.Builder}.
+   */
+  @Deprecated
+  public DrillClient(DrillConfig config, ClusterCoordinator coordinator) {
     this(config, coordinator, null, false);
   }
 
-  public DrillClient(DrillConfig config, ClusterCoordinator coordinator, boolean isDirect)
-    throws OutOfMemoryException {
+  /**
+   * @deprecated Create a DrillClient using {@link DrillClient.Builder}.
+   */
+  @Deprecated
+  public DrillClient(DrillConfig config, ClusterCoordinator coordinator, boolean isDirect) {
     this(config, coordinator, null, isDirect);
   }
 
-  public DrillClient(DrillConfig config, ClusterCoordinator coordinator, BufferAllocator allocator)
-      throws OutOfMemoryException {
+  /**
+   * @deprecated Create a DrillClient using {@link DrillClient.Builder}.
+   */
+  @Deprecated
+  public DrillClient(DrillConfig config, ClusterCoordinator coordinator, BufferAllocator allocator) {
     this(config, coordinator, allocator, false);
   }
 
+  /**
+   * @deprecated Create a DrillClient using {@link DrillClient.Builder}.
+   */
+  @Deprecated
   public DrillClient(DrillConfig config, ClusterCoordinator coordinator, BufferAllocator allocator, boolean isDirect) {
-    // if isDirect is true, the client will connect directly to the drillbit instead of
-    // going thru the zookeeper
-    this.isDirectConnection = isDirect;
-    this.ownsZkConnection = coordinator == null && !isDirect;
-    this.ownsAllocator = allocator == null;
-    this.allocator = ownsAllocator ? RootAllocatorFactory.newRoot(config) : allocator;
-    this.config = config;
-    this.clusterCoordinator = coordinator;
+    this(config,
+        allocator == null,
+        allocator == null ? RootAllocatorFactory.newRoot(config) : allocator,
+        coordinator == null && !isDirect,
+        coordinator,
+        isDirect,
+        true,
+        null,
+        true,
+        null,
+        config.getBoolean(ExecConstants.CLIENT_SUPPORT_COMPLEX_TYPES)
+    );
+  }
+
+  // used by DrillClient.Builder
+  protected DrillClient(DrillConfig config, boolean ownsAllocator, BufferAllocator allocator,
+                        boolean ownsZkConnection, ClusterCoordinator clusterCoordinator, boolean isDirectConnection,
+                        boolean ownsEventLoopGroup, EventLoopGroup eventLoopGroup, boolean ownsExecutor,
+                        ExecutorService executor, boolean supportComplexTypes) {
+    this.config = checkNotNull(config);
+    this.ownsAllocator = ownsAllocator;
+    this.allocator = checkNotNull(allocator);
+    this.ownsZkConnection = ownsZkConnection;
+    this.clusterCoordinator = clusterCoordinator;
+    this.isDirectConnection = isDirectConnection;
+    this.ownsEventLoopGroup = ownsEventLoopGroup;
+    this.eventLoopGroup = eventLoopGroup;
+    this.ownsExecutor = ownsExecutor;
+    this.executor = executor;
+    this.supportComplexTypes = supportComplexTypes;
```

Comment:
Nice work here. Would not it be nice if the builder provided "meaningful" defaults so that the complex creational patterns in this class particularly in #connect(...) could be eliminated further? For instance, why would executor ever be null or cluster coordinator be created later in the lifetime of this instance given instantiation cost is very minimal?